cmd/soroban-rpc/internal/config/config.go

```diff
@@ -1,39 +1,162 @@
 package config
 
 import (
+	"os"
+	"reflect"
 	"time"
 
 	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+
+	support "github.com/stellar/go/support/config"
+	"github.com/stellar/go/support/errors"
 )
 
-type LogFormat int
+// Config represents the configuration of a soroban-rpc server
+type Config struct {
+	ConfigPath string
 
-const (
-	LogFormatText = iota
-	LogFormatJSON
-)
+	Strict bool
+
+	StellarCoreURL         string
+	CaptiveCoreUseDB       bool
+	CaptiveCoreStoragePath string
+	StellarCoreBinaryPath  string
+	CaptiveCoreConfigPath  string
+	CaptiveCoreHTTPPort    uint
 
-type LocalConfig struct {
-	StellarCoreURL                   string
+	Endpoint                         string
+	AdminEndpoint                    string
+	CheckpointFrequency              uint32
 	CoreRequestTimeout               time.Duration
-	StellarCoreBinaryPath            string
-	CaptiveCoreConfigPath            string
-	CaptiveCoreStoragePath           string
-	CaptiveCoreHTTPPort              uint16
-	CaptiveCoreUseDB                 bool
+	DefaultEventsLimit               uint
+	EventLedgerRetentionWindow       uint32
 	FriendbotURL                     string
-	NetworkPassphrase                string
 	HistoryArchiveURLs               []string
-	LogLevel                         logrus.Level
-	LogFormat                        LogFormat
-	SQLiteDBPath                     string
 	IngestionTimeout                 time.Duration
-	EventLedgerRetentionWindow       uint32
-	TransactionLedgerRetentionWindow uint32
-	CheckpointFrequency              uint32
+	LogFormat                        LogFormat
+	LogLevel                         logrus.Level
 	MaxEventsLimit                   uint
-	DefaultEventsLimit               uint
 	MaxHealthyLedgerLatency          time.Duration
+	NetworkPassphrase                string
 	PreflightWorkerCount             uint
 	PreflightWorkerQueueSize         uint
+	SQLiteDBPath                     string
+	TransactionLedgerRetentionWindow uint32
+
+	// We memoize these, so they bind to viper flags correctly
+	optionsCache *ConfigOptions
+	flagsCache   *support.ConfigOptions
+}
+
+func (cfg *Config) Init(cmd *cobra.Command) error {
+	return cfg.flags().Init(cmd)
+}
+
+// We start with the defaults
+func (cfg *Config) SetValues() error {
+	if err := cfg.LoadDefaults(); err != nil {
+		return err
+	}
+
+	// Then we load from the cli flags and environment variables
+	if err := cfg.LoadFlags(); err != nil {
+		return err
+	}
+
+	// If we specified a config file, we load that
+	if cfg.ConfigPath != "" {
+		fileConfig, err := LoadConfigPath(cfg.ConfigPath, cfg.Strict)
```

Comment:
I tried that in an earlier iteration, but wasn't sure if it was clearer. Let me give it another go.

Reply:
> As is, it wouldn't merge fields that happen to have the default value.

Could you suggest an example test case that would show what you mean here?

Reply:
of course. Let's say that we have a Config variable named LogLevel. The default log level is 3. If the user would specify LOG_LEVEL = 0 in the toml file, that value would not be updated, since it's being considered to be the "default" value.